### ISSUE TITLE
no tradebot with Kraken

### ISSUE THREAD

**Issue by Kirbylix** [askmike/gekko#1176]
**Note: for support questions, please join our [Discord server](https://discord.gg/26wMygt)**

* **I'm submitting a ...**
[X ] bug report
[ ] feature request
[ ] question about the decisions made in the repository

* **Action taken** (what you did)
Start a new live Gekko -> Tradebot -> @Kraken EUR/XBT -> MACD

* **Expected result** (what you hoped would happen)
a Tradebot would start

* **Actual result** (unexpected outcome)
a market guardian was launched, which hangs in the state of loading + Dub #1090 

* **Other information** (e.g. detailed explanation, stacktraces, related issues, suggestions how to fix, links for us to have context, eg. stackoverflow, etc)
Using Gekko v0.5.2 and Gekko UI v0.2.0.
i use the current sample-config.js
Browser Console Log: 
Uncaught TypeError: Cannot read property 'toLowerCase' of undefined
    at Oe.e.$emit (build.js:6)
    at a.s.onmessage (build.js:6)
    at HTMLDivElement.<anonymous> (reconnecting-websocket.min.js:2)
    at WebSocket.h.onmessage (reconnecting-websocket.min.js:2)

Uncaught TypeError: Cannot read property 'start' of undefined
    at s.gekkoConfig (build.js:16)
    at Ri.get (build.js:6)
    at Ri.evaluate (build.js:6)
    at s.gekkoConfig (build.js:6)
    at startGekko (build.js:16)
    at s.n [as startGekko] (build.js:6)
    at start (build.js:16)
    at Proxy.n (build.js:6)
    at click (build.js:16)
    at HTMLAnchorElement.e (build.js:6)

**Comment by cmroche** [comment]
This is related to my change, ill look at it tonight.

**Comment by Kirbylix** [comment]
by the way, I ran a PaperTrader via console with the same settings as on top, and get the following error message: 
    `(ERROR):	error getting trades Error: Service:Unavailable
    at /home/pi/gekko/node_modules/kraken-api-es5/index.js:59:13
    at Generator.next (<anonymous>)
    at step (/home/pi/gekko/node_modules/kraken-api-es5/index.js:1:253)
    at /home/pi/gekko/node_modules/kraken-api-es5/index.js:1:423
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)`

and

`(ERROR):	error getting trades Error: API:Rate limit exceeded
    at /home/pi/gekko/node_modules/kraken-api-es5/index.js:59:13
    at Generator.next (<anonymous>)
    at step (/home/pi/gekko/node_modules/kraken-api-es5/index.js:1:253)
    at /home/pi/gekko/node_modules/kraken-api-es5/index.js:1:423
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)`

**Comment by cmroche** [comment]
These are not likely related to any changes, but instead that krakens API is notoriously unreliable in recent times.

**Comment by cmroche** [comment]
@Kirbylix After looking at the issue, actually I no longer think this is related to my changes to the kraken exchange. Instead the cause looks to be the config of fees

https://github.com/askmike/gekko/issues/1173
https://github.com/askmike/gekko/issues/1090

**Comment by Kirbylix** [comment]
when i click on the "Start" Button, i get this:
`Uncaught TypeError: Cannot read property 'start' of undefined
    at s.gekkoConfig (build.js:16)
    at Ri.get (build.js:6)
    at Ri.evaluate (build.js:6)
    at s.gekkoConfig (build.js:6)
    at startGekko (build.js:16)
    at s.n [as startGekko] (build.js:6)
    at start (build.js:16)
    at Proxy.n (build.js:6)
    at click (build.js:16)
    at HTMLAnchorElement.e (build.js:6)`

**Comment by pprados** [comment]
Hi, 
After download the last stable branch, I try to run 'node gekko/gekko.js --config sample-config.js'
I can receive a trace when the code call MACD.update() and MACD.check().
But, If i change the config.watch in sample-config.js:
exchange: 'poloniex', currency: 'USDT', asset: 'BTC',
changed to
exchange: 'kraken', currency: 'EUR', asset: 'XBT',
The methods update() or check() was not called.

I try with UI interface, it's the same.

**Comment by Kirbylix** [comment]
and another error with kraken:
`error getting trades { HTTPError: Response code 522 (undefined)
    at stream.catch.then.data (/home/pi/gekko/node_modules/got/index.js:182:13)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
  name: 'HTTPError',
  host: 'api.kraken.com',
  hostname: 'api.kraken.com',
  method: 'POST',
  path: '/0/public/Trades',
  protocol: 'https:',
  url: 'https://api.kraken.com/0/public/Trades',
  statusCode: 522,
  statusMessage: undefined,
  headers: 
   { date: 'Thu, 05 Oct 2017 16:42:19 GMT',
     'content-type': 'text/html; charset=UTF-8',
     'transfer-encoding': 'chunked',
     connection: 'close',
     'set-cookie': 
      [ '__cfduid=d819f08b00a98b9fe3192141bf021688d1507221724; expires=Fri, 05-Oct-18 16:42:04 GMT; path=/; domain=.kraken.com; HttpOnly' ],
     expires: 'Thu, 01 Jan 1970 00:00:01 GMT',
     'cache-control': 'no-store, no-cache, must-revalidate, post-check=0, pre-check=0',
     pragma: 'no-cache',
     'x-frame-options': 'SAMEORIGIN',
     server: 'cloudflare-nginx',
     'cf-ray': '3a91d4ff0b0d1ab7-DUS' } }`

**Comment by cmroche** [comment]
@Kirbylix that error means that cloud flare was unable to connect to kraken, again this is a problem with their API not being reliable, there won’t be much we can do for this aside from maybe reformatting the error message to look nicer.

**Comment by Kirbylix** [comment]
thanks for the info.
my current problem is that I have been running the bot for 4 days now (paper trader to see how it reacts). On Kraken I see the signals, but the bot doesn't say... neither in the console...

Config:
watch: kraken EUR-XBT
tradingAdvisor = true / MACD
paperTrader = true
performanceAnalyzer = true
trader = false
adviceLogger = true
candleWriter = false
adviceWriter = false

or did I do something wrong anyway?

**Comment by cmroche** [comment]
A changé was merged into develop late last week to fix a problem with paper trading and live trading not generating signals. Which branch are you using and when did you last pull?

**Comment by Kirbylix** [comment]
i have:
`git clone git://github.com/askmike/gekko.git`
=> master branche (stable)
the last pull was today

**Comment by cmroche** [comment]
The fix is not yet merged into stable.

**Comment by Kirbylix** [comment]
then i will try the develop branch, thx

**Comment by eZ3KIEL29** [comment]
same issue:
Gekko version: v0.5.8
Nodejs version: v8.6.0
sqlite 3.16.2

after a real buy or sell, i have this error.

Trader Received advice to go long. Buying  ETH
(INFO):     Attempting to BUY XXX ETH at kraken price: 264.24000
(ERROR):    no data

**Comment by eZ3KIEL29** [comment]
i have also import the newest kraken.js into gekko/exchanges from dev branch to solve a problem to get data from kraken.

**Comment by cmroche** [comment]
#1192 increases time to retry commands ads improvements to logging of no data error. This might fix the problem if you are hitting the Kraken API rate limit on retries.

**Comment by Kirbylix** [comment]
I got another ERROR:
`(ERROR):	Wanted to buy XBT but the amount is too small (0.024030249022) at kraken`
what is the minimum value?

According to kraken, their minimum is 0.002 BTC
[Kraken](https://support.kraken.com/hc/en-us/articles/205893708-What-is-the-minimum-order-size-)

**Comment by Kirbylix** [comment]
All right, I've been running the tradebot for almost five days now. I have received messages via telegram for trades. But the bot didn't buy or sell anything in time...

could someone please upload a working config. js for a tradebot?

**Comment by cmroche** [comment]
@Kirbylix There have been a few bug fixed merged to kraken, one of which deals with precision issues in orders which was merged yesterday. Please grab the latest code on the develop branch and try again, if you continue to get new errors post them here.

**Comment by Kirbylix** [comment]
I pulled the develop branch. 
if there are any errors, I'll let you know.

**Comment by metodi-yanev** [comment]
Hey folks, i've tried the bot with the latest changes, and it still enters a 'no data' - retry loop. What's changed is that it logs the error from the 'getPortfolio' method - 'no data (getPorfolio)'. Actually when i start new live gekko it starts with the same error `Kraken did not return portfolio for XBT, assuming 0.`. Any ideas? (Command line user)

**Comment by xerix32** [comment]
I've found the issue (work for me) introduced with "roundAmount" and new kraken precision (in kraken.js exchange).

Try to change 
`- var market = this.getCapabilities().markets.find(function(market){ return market.pair[0] === this.currency && market.pair[1] === this.asset });`

with
```
 +  var parent = this;
 +  var market = Trader.getCapabilities().markets.find(function(market){ return market.pair[0] === parent.currency && market.pair[1] === parent.asset });
```
and try if works for you.

The message "no data" was misleading, and overwrite the true problem (variable scope error), I added more "talking" debug info :
`+ log.debug('[kraken.js] entering "setBalance" callback after kraken-api call, err:', err, ' data:' , data);`
before overwriting
`err = "no data";`
that helps me "digging the bug".


I "commited" a pull request
[pull request](https://github.com/askmike/gekko/pull/1229/)
